avbroot/src/crypto.rs

```diff
@@ -110,6 +161,169 @@ impl PassphraseSource {
     }
 }
 
+fn check_key_size(size: usize) -> Result<()> {
+    // RustCrypto does not support 8192-bit keys.
+    if size > 4096 / 8 {
+        return Err(Error::UnsupportedKey(size));
+    }
+
+    Ok(())
+}
+
+#[derive(Clone)]
+pub enum RsaSigningKey {
+    Internal(RsaPrivateKey),
+    External {
+        program: PathBuf,
+        public_key_file: PathBuf,
+        public_key: RsaPublicKey,
+        passphrase_source: PassphraseSource,
+    },
+}
+
+impl RsaSigningKey {
+    /// Size of key in bytes.
+    pub fn size(&self) -> usize {
+        match self {
+            Self::Internal(key) => key.size(),
+            Self::External { public_key, .. } => public_key.size(),
+        }
+    }
+
+    /// Get the public key portion of the signing key.
+    pub fn to_public_key(&self) -> RsaPublicKey {
+        match self {
+            RsaSigningKey::Internal(key) => key.to_public_key(),
+            RsaSigningKey::External { public_key, .. } => public_key.clone(),
+        }
+    }
+
+    /// Sign the digest with the specified signature algorithm.
+    pub fn sign(&self, algo: SignatureAlgorithm, digest: &[u8]) -> Result<Vec<u8>> {
+        if digest.len() != algo.digest_len() {
+            return Err(Error::InvalidDigestLength(digest.len(), algo));
+        }
+
+        check_key_size(self.size())?;
+
+        match self {
+            Self::Internal(key) => {
+                match algo {
+                    // We don't support signing with insecure algorithms.
+                    SignatureAlgorithm::Sha1WithRsa => Err(Error::UnsupportedAlgorithm(algo)),
+                    SignatureAlgorithm::Sha256WithRsa => {
+                        let scheme = Pkcs1v15Sign::new::<Sha256>();
+                        Ok(key.sign(scheme, digest).map_err(Error::RsaSign)?)
+                    }
+                    SignatureAlgorithm::Sha512WithRsa => {
+                        let scheme = Pkcs1v15Sign::new::<Sha512>();
+                        Ok(key.sign(scheme, digest).map_err(Error::RsaSign)?)
+                    }
+                }
+            }
+            Self::External {
+                program,
+                public_key,
+                public_key_file,
+                passphrase_source,
+            } => {
+                let key_size = public_key.size();
```

Comment:
[`.size()` returns in bytes](https://docs.rs/rsa/latest/rsa/struct.RsaPublicKey.html#method.size) instead of bits, so I'm getting `SHA256_RSA256` with an RSA 2048 key here.